Drop support for Python 2.6

jdufresne opened the pull request:
I would humbly like to suggest Pillow drop support for Python 2.6.

Python 2.6 has had its last release and been end of life for over [3 years](https://www.python.org/dev/peps/pep-0361/). By removing support, Pillow can move forward with the greater Python community by adopting newer language features and be more forward compatible. pip, the Python package manager, has recently announced they are [dropping 2.6 support](https://mail.python.org/pipermail/distutils-sig/2016-November/029785.html).

I've tried my best to remove as much 2.6 specific code as I can, including the `'Programming Language :: Python :: 2.6'` trove classifier from `setup.py`.

Interestingly, there are some existing uses of `str.format()` [automatic](https://github.com/jdufresne/Pillow/blob/5aeb0ed9727a8c51c98d11788c0135c54e88b1b7/PIL/IcnsImagePlugin.py#L317) [numbering](https://github.com/jdufresne/Pillow/blob/5aeb0ed9727a8c51c98d11788c0135c54e88b1b7/Tests/test_image_resample.py#L205) in the code. Those parts are already incompatible with 2.6.

Thanks for considering.

Code changes in this PR include: 

* Drop unittest2 requirement
* Use set literals
* Use dict/set comprehension
* Use `str.format()` automatic numbering

hugovk commented:
See https://github.com/python-pillow/Pillow/issues/1637 for some stats form January this year.

A quick check of PyPI downloads for Pillow's last release, 3.4.2 made 18 days ago, suggests only 0.68% (1,401/204,807) were for a Python 2.6 package.

\*(`vanity pillow -p 3\.4\.2` and `vanity pillow -p 3\.4\.2.*2[\.]?6`)

hugovk commented:
https://gist.github.com/juanpabloaj/dffc6900f80abcfe8ce121a39cffa743 has more recent stats:

> # Total of pip packages downloaded separated by Python versions
>
> From June 26, 2016 (python 3.5.2 release) to Aug. 31, 2016.
>
> Python versions from 2.6 to 3.5
> ![Python versions from 2.6 to 3.5](https://cloud.githubusercontent.com/assets/1324225/20022640/20a75910-a2e4-11e6-99cc-b3c8fbb159b2.png)
>
> Without 2.7
> 
![Without 2.7](https://cloud.githubusercontent.com/assets/1324225/20022653/384e047e-a2e4-11e6-8cf1-646a6ecb7c1e.png)
...
> Here's a log scale version of the same data.
> 
![Here's a log scale version of the same data](https://cloud.githubusercontent.com/assets/1324225/20022671/506e5da6-a2e4-11e6-9020-f7e97916a79d.png)

wiredfool commented:
Not sure where the 3.1 data comes from, since we don't actually produce a binary for it and IIRC we've never been compatible with it. 

I'd be willing to have a discussion on dropping 3.2 and 2.6. They're both ancient, and they both are significant additions to the test matrix in terms of platforms that break. 

From what I can tell, 2.6 isn't shipped anywhere, and is only in use by people who don't care about updates, or can't update for some external reason. They're implicitly or explicitly taking on backporting duties for a whole bunch of other things so, I'd be willing to push the job down to them. (unless @manisandro objects, as the only possible people shipping 2.6 would be with some enterprise supported RH type product).

3.2 was shipped with a currently supported Ubuntu LTS release (12.04), but clearly not many people are downloading from pypi on that platform.

manisandro commented:
@wiredfool No objections here (though I only can speak for Fedora side since I'm not involved with RHEL).

hugovk commented:
The graphs on this page that include 3.1 are for all PyPI packages, not just Pillow.

wiredfool commented:
Ah, thanks @hugovk.  If that's the case, then we probably account for some small percentage of those downloads due to pip installing packages when running our 3.2 ci.

hugovk commented:
Yep, I bet CIs in general account for a large number of 2.6 and 3.2 downloads.

I'd also be fine with removing 2.6 and especially 3.2. If it's trivial to support, it doesn't matter. But definitely 3.2 is the most troublesome and more and more libraries are dropping 3.2 support, soon it will be harder for us too. 

2.6 doesn't seem that difficult to support (at the moment), still a few edge cases. But if it's not being used much, and no longer supported, there's less point in even doing so.

Ping also @python-pillow/pillow-team.

aclark4life commented:
👍 2.7 should probably be the only 2.x we support at this point.

homm commented:
> I'd be willing to have a discussion on dropping 3.2 and 2.6. 

👍

hugovk commented:
Here's some stats of Pillow downloads, for 3.2.0 (released 1st April and the most recent release to be out the full three months before the next release) and the most recent 3.4.2 (released 18 days ago).

Pillow 3.2.0
![image](https://cloud.githubusercontent.com/assets/1324225/20030058/b5f60bac-a365-11e6-8e36-bf0065fd732a.png)

Pillow 3.4.2
![image](https://cloud.githubusercontent.com/assets/1324225/20030056/a6df82ce-a365-11e6-8d34-13d545d7d6c5.png)

![image](https://cloud.githubusercontent.com/assets/1324225/20030091/83a5616a-a366-11e6-83e2-4630ef835963.png)

Notes:
 * These percentages are excluding source-only packages. If source were included, source alone accounts for 87% for Pillow 3.2.0 and 74% for Pillow 3.4.2.
 * Python 3.2 has gone to zero for 3.4.2 because we don't make binary packages for that any more.

homm commented:
I believe than 90% of Python 2.6 users is CIs

jdufresne commented:
Would you like me to update this PR/commit to drop 3.2 at the same time? I have no problem making that change. Thanks.

aclark4life commented:
@jdufresne Unless someone objects, I'd say yes let's get 2.6 & 3.2 removal done in one PR

jdufresne commented:
> Unless someone objects, I'd say yes let's get 2.6 & 3.2 removal done in one PR

Done in the updated commit. All additional feedback welcome. Thanks.

hugovk commented on docs/installation.rst:
```diff
@@ -340,44 +342,40 @@ current versions of Linux, macOS, and Windows.
 +----------------------------------+-------------+------------------------------+--------------------------------+-----------------------+
-| Windows 8.1 Pro                  |Yes          | 2.6,2.7,3.2,3.3,3.4          | 2.4.0                          |x86,x86-64             |
+| Windows 8.1 Pro                  |Yes          | 2.7,3.3,3.4                  | 2.4.0                          |x86,x86-64             |
 +----------------------------------+-------------+------------------------------+--------------------------------+-----------------------+
```
Might it still be useful leaving the historical version support here? If someone still uses Python 2.6 or 3.2, they can still see here what versions were good on what platforms. Of course, we won't bother testing 2.6 or 3.2 for future ones.

hugovk commented:
It might be worth bumping the next Pillow release to 4.0.0 to make it abundantly clear it's not backwards compatible with Python 2.6 and 3.2. 

It'd follow suit with Pillow 2.0.0 dropping Python versions 2.4 and 2.5, and also be much easier to remember.

Thoughts?

jdufresne commented:
> 

Might it still be useful leaving the historical version support here? If someone still uses Python 2.6 or 3.2, they can still see here what versions were good on what platforms. Of course, we won't bother testing 2.6 or 3.2 for future ones.

Reverted this doc change in the latest version.

> It might be worth bumping the next Pillow release to 4.0.0 to make it abundantly clear it's not backwards compatible with Python 2.6 and 3.2. 

I agree. I think it helps make it more clear to library consumers. I wasn't sure what the exact version policy was when I originally wrote the change. I have made this change in the latest version. If there is any disagreement about changing the version, let me know and I can revert it.

homm commented:
1. @jdufresne It would be better to remove 2.6 and 3.2 at least in two separate commits.
2. A good practice is not to modify changelog and versions in Pull requests. This information not related to code changes itself, it is more an administrative question. It's more conveniently to see version changes as separate commits in master. Also, as you can see, the conflicts probability is high.
3. `3.0` doesn't introduce significant changes or backward incompatibilities. In [the previous discussion](https://github.com/python-pillow/Pillow/issues/1087) about versioning policy, there was a decision to do "aggressive but fair deprecation" not tied to version numbers. Also, was decided to remove feature after one year after depreciation. I suppose we must adhere to the decisions made or change decisions in the more explicit way.

wiredfool commented on depends/ubuntu_12.04.sh:
```diff
@@ -2,7 +2,7 @@
 
 #
 # Installs all of the dependencies for Pillow for Ubuntu 12.04
-# for both system Pythons 2.7 and 3.2
```
This is incorrect, as the system python3 is version 3.2

wiredfool commented on appveyor.yml:
```diff
@@ -1,68 +1,68 @@
-version: 3.5.pre.{build}
```
Please don't bulk change line endings.

jdufresne commented:
Integrated feedback in latest version. All additional feedback is welcome.

>  It would be better to remove 2.6 and 3.2 at least in two separate commits.

Split changes into two commits.

> A good practice is not to modify changelog and versions in Pull requests.

I have removed the changes to `CHANGES.rst` and the version bump. I'll leave these as administrative tasks.

> This is incorrect, as the system python3 is version 3.2

Reverted this change.

> Please don't bulk change line endings.

Reverted this changed (was only related to the version bump)

jdufresne commented:
> 3.0 doesn't introduce significant changes or backward incompatibilities. In the previous discussion about versioning policy, there was a decision to do "aggressive but fair deprecation" not tied to version numbers. Also, was decided to remove feature after one year after depreciation. I suppose we must adhere to the decisions made or change decisions in the more explicit way.

Sorry, it isn't clear if this is asking me to make an explicit modification to my PR to make it more merge friendly for the project. If you are requesting a specific change, can you please state it explicitly?

aclark4life commented:
@jdufresne Hang tight, I think @homm voiced concern about @hugovk's suggestion about bumping to 4.0.0. While I tend to like to use "big" version number changes as @hugovk suggests (as an indicator of some significant change), @homm's point about adhering to our own policies is valid.

Aside from the ticket that @homm referenced, do we have our policies written down somewhere (I assume so, but if not, we should)? I'm inclined to not worry too much about it and just drop them and move on, but let's see if we can make everyone happy by (re)making decisions about when to drop old features and what our version number policy is before we move forward.

wiredfool commented:
This post still reflects my thinking: https://github.com/python-pillow/Pillow/issues/1087#issuecomment-121655443

For this particular PR, we're not in a place where the packagers really matter anymore, as either they're on newer python or they're only backporting bugfixes. So, if they're maintaining an old python, they're not maintaining our compatibility with that old python for those CVE bugfixes which affect them.

jdufresne commented:
> This is also an invalid change. This is reporting a test result, which was with the system python3 on ubuntu 12.04

Thanks for the review. Fixed in latest version.

wiredfool commented:
Meged in #2192

hugovk commented:
This needs to go in the release notes as well.
https://github.com/python-pillow/Pillow/tree/master/docs/releasenotes

hugovk commented:
Bikeshed discussion about version number: https://github.com/python-pillow/Pillow/issues/2244
